src/app/region.js

```diff
@@ -0,0 +1,188 @@
+  })
+  // Coordinates have array of lines
+  // Lines have array of points
+  // Points have one array of latlngs [lng, lat]
+  for (let lineIndex = coordinates.length - 1; lineIndex >= 0; lineIndex--) {
+    const line = coordinates[lineIndex]
+    for (let pointsIndex = line.length - 1; pointsIndex >= 0; pointsIndex--) {
+      const points = line[pointsIndex]
+      for (let coordIndex = points.length - 1; coordIndex >= 0; coordIndex--) {
+        const point = points[coordIndex]
+        const lat = point[1]
+        const lng = point[0]
+        // Checking if latlng is within bounding box
+        // If not remove from points
+        if (lng < Number(bounds.west) - buffer || lng > Number(bounds.east) + buffer || lat < Number(bounds.south) - buffer || lat > Number(bounds.north) + buffer) {
+          points.splice(coordIndex, 1)
+        }
+      }
+      // If no points in line, remove from line
+      if (points.length === 0) { line.splice(pointsIndex, 1) }
+    }
+    // If no lines in coordinates, remove from coordinates
+    if (line.length === 0) { coordinates.splice(lineIndex, 1) }
+  }
+
+  // If no coordinates, remove entire feature from array of features
+  for (let i = features.length - 1; i >= 0; i--) {
+    const feature = features[i]
+    if (feature.geometry.coordinates.length === 0) { features.splice(i, 1) }
+  }
+}
+
+function getBboxArea (bounds) {
+  const width = (bounds.east - bounds.west)
+  const height = (bounds.north - bounds.south)
+  const area = width * height
+  return area
+}
+
+export function showRegion (bounds) {
+  // If bounds are cleared, remove data source from tangram
+  if (!bounds) {
+    const scene = getCurrentScene()
+    delete scene.sources.routes
+    setCurrentScene(scene)
+    return
+  }
+
+  // If area of bounding box exceeds max_area, display error
+  // Arbitrary number for max_area, edit if needed
+  const maxArea = 0.01
+  const area = getBboxArea(bounds)
+  if (area > maxArea) {
+    const message = 'Please zoom in and reduce the size of your bounding box'
+    store.dispatch(setRouteError(message))
+    return
+  }
+
+  // First, convert bounds to waypoints
+  // This way we can get the bounding box from valhalla and suffixes
+  const waypoints = [L.latLng(bounds.south, bounds.west), L.latLng(bounds.north, bounds.east)]
+  // Go get route using Valhalla
+  getRoute(host, waypoints)
+    // After getting route, get the bounding box and the relevant tiles from box
+    .then((response) => {
+      const bbox = response.trip.summary
+      const suffixes = getSuffixes(bbox)
+      return suffixes
+    })
+    // Second, fetch the OSMLR Geometry tiles using the suffixes
+    .then((suffixes) => {
+      store.dispatch(startLoading())
+      fetchOSMLRGeometryTiles(suffixes)
+        .then((results) => {
+          const features = results.features
+          // Remove from geojson, routes outside bounding box (bounds)
+          withinBbox(features, bounds)
```

Comment:
Something like that, but let me try to explain better. You might know some of this already but it's easier for me if I set it up.

A side effect is when a function does some work on something outside of the function:

```js
const myObj = {}

function doStuff() {
   // myObj is never passed in, it just accesses it from outside the function and modifies it
   myObj.foo = 'bar'
}

doStuff()
console.log(myObj) // -> { foo: "bar" }
```

So then you pass in `myObj` as an argument:

```js
const myObj = {}

function doStuff(obj) {
   obj.foo = 'bar'
}

doStuff(myObj)
console.log(myObj) // -> { foo: "bar" }
```

This helps a little, but the problem is that when the function `doStuff()` does stuff, it doesn't `return` anything. Instead it relies on objects being passed-in-by-reference to modify the original `myObj`. So it's still a side effect.

Suppose `doStuff()` is a library function that gets imported so it's hard to see what it actually does. So when I see this code:

```js
import {doStuff} from 'some-library'
const myObj = {}
doStuff(myObj)
// WHAT DID IT DO?!
```

I see a function being called, but... then what? It's a super black box.

Ideally I want to see something like this:

```js
import {doStuff} from 'some-library'
const myObj = {}
const myObjWithStuffDone = doStuff(myObj)
// oh ok, I know it returns something
```

And `doStuff()` might look like this:

```js
function doStuff(obj) {
   obj.foo = 'bar'
   // I did work to `obj`, let's return it!
   return obj
}
```

So it's a little more obvious how the control flow works for one variable to go somewhere. But the implementation of `doStuff()` _still_ has a side effect, it's just less obvious now! (Even though we return the value from `doStuff` and assign it to a new variable `myObjWithStuffDone`, the reality is that the original `myObj` is _also_ modified.

```js
const myObj = {}
const myObjWithStuffDone = doStuff(myObj)

console.log(myObjWithStuffDone) // -> { foo: 'bar' } (I expected this!)
console.log(myObj) // -> { foo: 'bar' } (Huh!)
```

In the ideal scenario, `doStuff()` would also want to be a pure function instead of having any side effects. It might be implemented this way:

```js
function doStuff(obj) {
    const newObj = Object.assign({}, obj)
    newObj.foo = 'bar'
    return newObj
}

const myObj = {}
const myNewObj = doStuff(myObj)

console.log(myObj) // -> {}
console.log(myNewObj) // -> { foo: 'bar' }
```

Now it has no side effects at all. The original `myObj` is cloned inside of `doStuff`, and a new result is returned.

This is very similar to that `array.reverse()` situation we ran into earlier this week when modifying the array using one of these array methods to make it display correctly in the Legend actually made the routes display the wrong colors. So whenever we can, we want to (a) avoid side effects inside of functions (b) return values from functions (c) assign the return values of functions to a variable.

Using `const` everywhere does meant we need to set up a lot of new variables (which is okay), but if you have a lot of intermediary variables you don't mind overwriting, it's also perfectly acceptable to set up a `let` instead.

```js
let myObj = {}
myObj = doStuff(myObj)
myObj = doOtherStuff(myObj)
// etc
```

Hope that helps!